nestedtensor/nested/masking.py

```diff
@@ -1,111 +1,184 @@
-"""
-def nested_tensor_from_tensor_mask(tensor, mask, nested_dim=None):
-    # TODO: Should be possible with views only.
-    if not mask.dim() > 0:
-        raise RuntimeError("Mask has to have dimention > 0")
-
-    # TODO: Need to define empty-list and 0 numel semantics
-    # TODO: Test edge-cases and kwargs
-    def _merge(tensors, nested_dim=None):
-        assert len(tensors) > 0
-
-        def _mergable_tensors(tensors, nested_dim):
-            if nested_dim is None:
-                size0 = tensors[0].size()
-                compliant = all(torch.is_tensor(t) for t in tensors)
-                compliant = all(size0 == t.size()
-                                for t in tensors) and compliant
-                return compliant
-            else:
-                return nested_dim < 1
 
-        if _mergable_tensors(tensors, nested_dim):
-            return torch.stack(tensors)
-        else:
-            return creation.nested_tensor(tensors)
+# Constructs nested tensor from passed tensor and mask.
+def nested_tensor_from_tensor_mask(tensor, mask, nested_dim=1):
+    if tensor is None:
+        raise RuntimeError("Tensor can't be undefined (None).")
+
+    if mask is None:
+        raise RuntimeError("Mask can't be undefined (None).")
+
+    # Scalar was passed
+    if tensor.dim() == 0:
+        raise RuntimeError("Can't construct nested tensor from a scalar.")
+
+    if nested_dim == 0:
+        raise RuntimeError("Nested dimension can't be 0.")
 
-    if nested_dim is not None and mask.dim() < nested_dim:
-        raise ValueError("Mask dimension ({0}) is too small to construct nested tensor with nested dimension of {1}".format(mask.dim(), nested_dim))
+    if nested_dim is not None and nested_dim > tensor.dim():
+        raise RuntimeError("Nested dimension ({0}) can't be bigger than data tensor dimension ({1}).".format(nested_dim, tensor.dim()))
 
-    if mask.dim() == 1:
-        tensors = [tensor[i] if mask[i]
-                   else None for i in range(len(mask))]
-        tensors = list(filter(lambda x: x is not None, tensors))
+    if tensor.numel() == 0 and mask.numel() != 0:
+        raise RuntimeError("Data tensor can't be emtpy if a mask has values.")
+
+    if tensor.numel() != 0 and mask.numel() == 0:
+        raise RuntimeError("Mask tensor can't be emtpy if a data tensor has values.")
+
+    return nt_from_tensor_mask(tensor, mask, nested_dim, tensor.dtype)
+
+
+def nt_from_tensor_mask(tensor, mask, nested_dim, dt):
+    def _merge(tensors, nested_dim):
         if len(tensors) == 0:
-            return tensor.narrow(0, 0, 0)
-        return _merge(tensors, nested_dim)
+            return torch.tensor([], dtype=dt)
+        return torch.stack(tensors)
+
+    if nested_dim == 0:
+        if (mask.numel() == 0) or (mask.numel() == 1 and mask == True):
+            return tensor
+
+        if mask.dim() == 1:
+            tensors = [tensor[i] if mask[i] else None for i in range(len(mask))]
```

Comment:
You should be able to use mask as an index if its dim() 1. As in ```tensor[mask]```. For example

```
>>> import torch
>>> a = torch.randn(3, 4)
>>> b = torch.tensor([True, False, True])
>>> b
tensor([ True, False,  True])
>>> a[b]
tensor([[-0.3448, -1.0036, -0.2669,  1.5595],
        [ 0.2324, -0.5979,  0.8668, -0.0993]])
>>> a
tensor([[-0.3448, -1.0036, -0.2669,  1.5595],
        [ 0.6527,  0.4671,  0.8961, -0.3756],
        [ 0.2324, -0.5979,  0.8668, -0.0993]])
```

Reply:
it will return diufferent results in most of the cases. 
tensor([1, 2, 3, 4]) vs [tensor(1), tensor(2), tensor(3), tensor(4)]